website/content/configuration.md
+++
title = "Configuration"
template = "docgen.html"

[extra]
toc = true

# DO NOT EDIT
# Use `make_docs.sh` to generate content
+++
# Config file

Launch retherm with the path to your custom configuration.

```bash
retherm --config ./your_config.toml
```

All config options have a default; you only need to include options
you would like to override in your configuration file.

## temp_deadband

The temperature difference from the setpoint required to trigger an action.

For example, with a target heat temp of 20, and deadband set to 0.4,
the hvac system will turn heat on when temp drops to 19.6.

Defaults to 0.6

## temp_overrun

The temperature difference past the setpoint required to stop an action.

For example, with a target heat temp of 20, and overrun set to 0.2,
the hvac system will turn heat off when temp reaches 20.2.

Defaults to 0.4

## min_off_time

Minimum off time for cooling to allow AC refrigerant pressures to equalize.

Defaults to "5m"

## default_fan_timeout

Default amount of time to run fan, when fan mode is activated.

Defaults to "15m"

## storage_dir

Directory to store app state.

Defaults to "/media/data"

# Away Mode

```toml
[away_mode]
temp_heat = 16.0
temp_cool = 20.0
timeout = "0s"
```

## temp_heat

Away temp for heating mode, default 16.0

## temp_cool

Away temp for cooling mode, default 22.0

## timeout

Duration of no proximity movement before going into away mode,
or set to zero to disable away mode. Default "30m".

# Backplate

```toml
[backplate]
near_pir_threshold = 15
serial_port = "/dev/ttyO2"
wiring = { heat_wire: "W1", cool_wire: "Y1" }
```

## near_pir_threshold

Minimum near proximity value to be considered as movement, default 15

## serial_port

Path to backplate serial device file, default "/dev/ttyO2"

## wiring

HVAC wiring configuration, default `{ heat_wire: "W1", cool_wire: "Y1" }`.
Valid wire names: W1, Y1, G, OB, W2, Y2, Star.

# Home Assistant

```toml
[home_assistant]
friendly_name = "Hallway"
encryption_key = "..."
```

## object_id

Object ID used internall by home assistant.
Defaults to "climage.{node_name}".

## listen_addr

Listen address for ESP Home API server, default "0.0.0.0:6053"

## encryption_key

Encryption key as 32 byte base64 string. When not provided, the
connection uses plaintext messages.
See [ESP Home Native API](https://esphome.io/components/api/)
for a tool that generates a random key.

## server_info

Server info (not typically displayed in Home Assistant).
Defaults to "ReTherm {version}".

## node_name

Node name, defaults to the system hostname

## friendly_name

Friendly name displayed in as label for thermostat control

## manufacturer

Manufactuer name, defaults to "Nest"

## model

Model name, defaults to "Gen2 Thermostat"

## mac_address

Mac address, defaults to address of system interface address

# Backlight

```toml
[backlight]
brightness = 108
timeout = "15s"
```

## brightness

Screen brightness, defaults to 108 (max 120)

## timeout

Timeout before screen turns off, defaults to "15s"

# Schedule

```toml
[[schedule_heat]]
days_of_week = "EveryDay"
set_points = [
   { time = "08:00", temp = 20.0 },
   { time = "22:00", temp = 16.0 },
]
```

* Heating schedule `[[schedule_heat]]`
* Cooling schedule `[[schedule_cool]]`

You can define more than one schedule entry, and it will overlap the
previous. In the example below, the temperature will be set to 20.0
at 8am everyday, and set down to 16.0 at 9am Monday and Wednsday.

```toml
[[schedule_heat]]
days_of_week = "EveryDay"
set_points = [
   { time = "08:00", temp = 20.0 }
]

[[schedule_heat]]
days_of_week = ["Monday", "Wednsday"]
set_points = [
   { time = "09:00", temp = 16.0 }
]
```

## days_of_week

Days of the week.

One of "EveryDay", "WeekDays", "WeekEnd"

Or...

List of weekdays ["Monday", "Tuesday", ...]

## set_points

List of set points with time of day and temperature



src/config.rs
/*
 * ReTherm - Home Assistant native interface for Gen2 Nest thermostat
 * Copyright (C) 2026 Josh Kropf <josh@slashdev.ca>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use std::{fs, path::{Path, PathBuf}, time::Duration};

use anyhow::Result;
use serde::Deserialize;

mod config_de;
mod schedule_config;

pub use schedule_config::*;

use crate::{env, state::HvacMode};

/// Config file
///
/// Launch retherm with the path to your custom configuration.
///
/// ```bash
/// retherm --config ./your_config.toml
/// ```
///
/// All config options have a default; you only need to include options
/// you would like to override in your configuration file.
#[derive(Deserialize, Debug, Clone)]
#[serde(default)]
pub struct Config {
    /// The temperature difference from the setpoint required to trigger an action.
    ///
    /// For example, with a target heat temp of 20, and deadband set to 0.4,
    /// the hvac system will turn heat on when temp drops to 19.6.
    ///
    /// Defaults to 0.6
    pub temp_deadband: f32,

    /// The temperature difference past the setpoint required to stop an action.
    ///
    /// For example, with a target heat temp of 20, and overrun set to 0.2,
    /// the hvac system will turn heat off when temp reaches 20.2.
    ///
    /// Defaults to 0.4
    pub temp_overrun: f32,

    /// Minimum off time for cooling to allow AC refrigerant pressures to equalize.
    ///
    /// Defaults to "5m"
    #[serde(deserialize_with = "config_de::duration")]
    pub min_off_time: Duration,

    /// Path to file where retherm will store app state.
    ///
    /// Defaults to "/media/data/rether.state.toml"
    pub state_file_path: PathBuf,

    pub away_mode: AwayConfig,
    pub backplate: BackplateConfig,
    pub home_assistant: HomeAssistantConfig,
    pub backlight: BacklightConfig,
    pub schedule_heat: Vec<ScheduleConfig>,
    pub schedule_cool: Vec<ScheduleConfig>
}

impl Config {
    pub fn load<P: AsRef<Path>>(file_path: P) -> Result<Self> {
        let toml_src = fs::read_to_string(file_path)?;
        let config = toml::from_str(&toml_src)?;
        Ok(config)
    }

    pub fn schedule_for_mode(&self, mode: &HvacMode) -> Option<&[ScheduleConfig]> {
        match mode {
            HvacMode::Heat => {
                if self.schedule_heat.len() > 0 {
                    Some(&self.schedule_heat)
                } else {
                    None
                }
            }
            HvacMode::Cool => {
                if self.schedule_cool.len() > 0 {
                    Some(&self.schedule_cool)
                } else {
                    None
                }
            }
            _ => None
        }
    }
}

impl Default for Config {
    fn default() -> Self {
        Self {
            away_mode: AwayConfig::default(),
            backplate: BackplateConfig::default(),
            home_assistant: HomeAssistantConfig::default(),
            backlight: BacklightConfig::default(),
            schedule_heat: Vec::new(),
            schedule_cool: Vec::new(),
            temp_deadband: 0.6,
            temp_overrun: 0.4,
            min_off_time: Duration::from_mins(5),
            state_file_path: PathBuf::from("/media/data/retherm.state.toml"),
        }
    }
}

/// Home Assistant
///
/// ```toml
/// [home_assistant]
/// friendly_name = "Hallway"
/// encryption_key = "..."
/// ```
#[derive(Deserialize, Debug, Clone)]
#[serde(default)]
pub struct HomeAssistantConfig {
    /// Object ID used internall by home assistant.
    /// Defaults to "climage.{node_name}".
    pub object_id: Option<String>,

    /// Listen address for ESP Home API server, default "0.0.0.0:6053"
    pub listen_addr: String,

    /// Encryption key as 32 byte base64 string. When not provided, the
    /// connection uses plaintext messages.
    /// See [ESP Home Native API](https://esphome.io/components/api/)
    /// for a tool that generates a random key.
    pub encryption_key: Option<String>,

    /// Server info (not typically displayed in Home Assistant).
    /// Defaults to "ReTherm {version}".
    pub server_info: String,

    /// Node name, defaults to the system hostname
    pub node_name: Option<String>,

    /// Friendly name displayed in as label for thermostat control
    pub friendly_name: String,

    /// Manufactuer name, defaults to "Nest"
    pub manufacturer: String,

    /// Model name, defaults to "Gen2 Thermostat"
    pub model: String,

    /// Mac address, defaults to address of system interface address
    pub mac_address: Option<String>
}

impl HomeAssistantConfig {
    pub fn get_object_id(&self) -> String {
        if let Some(object_id) = &self.object_id {
            object_id.clone()
        } else {
            format!("climate.{}", self.get_node_name())
        }
    }

    pub fn get_node_name(&self) -> String {
        let pkg_name = env::get_pkg_name();

        if let Some(node_name) = &self.node_name {
            node_name.clone()
        } else {
            match env::get_hostname() {
                Ok(hostname) => hostname,
                Err(e) => {
                    log::error!("get_hostname: '{e}'; using '{pkg_name}'");
                    pkg_name.into()
                }
            }
        }
    }

    pub fn get_mac_address(&self) -> String {
        const FAKE_MAC: &str = "01:02:03:04:05:06";

        if let Some(mac_addr) = &self.mac_address {
            mac_addr.clone()
        } else {
            match env::get_mac_addr() {
                Ok(mac_addr) => {
                    if let Some(mac_addr) = mac_addr {
                        mac_addr
                    } else {
                        log::warn!("get_mac_addr None; using '{FAKE_MAC}'");
                        FAKE_MAC.into()
                    }
                }
                Err(e) => {
                    log::error!("get_mac_addr: '{e}'; using '{FAKE_MAC}'");
                    FAKE_MAC.into()
                }
            }
        }
    }
}

impl Default for HomeAssistantConfig {
    fn default() -> Self {
        Self {
            object_id: None,
            listen_addr: "0.0.0.0:6053".to_string(),
            encryption_key: None,
            server_info: format!("ReTherm {}", env::get_pkg_ver()),
            node_name: None,
            friendly_name: "ReTherm Thermostat".to_string(),
            manufacturer: "Nest".to_string(),
            model: "Gen2 Thermostat".to_string(),
            mac_address: None
        }
    }
}

/// Backlight
///
/// ```toml
/// [backlight]
/// brightness = 108
/// timeout = "15s"
/// ```
#[derive(Deserialize, Debug, Clone)]
#[serde(default)]
pub struct BacklightConfig {
    /// Screen brightness, defaults to 108 (max 120)
    pub brightness: u32,

    /// Timeout before screen turns off, defaults to "15s"
    #[serde(deserialize_with = "config_de::duration")]
    pub timeout: Duration
}

impl Default for BacklightConfig {
    fn default() -> Self {
        Self {
            brightness: 108,
            timeout: Duration::from_secs(15)
        }
    }
}


/// Away Mode
///
/// ```toml
/// [away_mode]
/// temp_heat = 16.0
/// temp_cool = 20.0
/// timeout = "0s"
/// ```
#[derive(Deserialize, Debug, Clone)]
#[serde(default)]
pub struct AwayConfig {
    /// Away temp for heating mode, default 16.0
    pub temp_heat: f32,

    /// Away temp for cooling mode, default 22.0
    pub temp_cool: f32,

    /// Duration of no proximity movement before going into away mode,
    /// or set to zero to disable away mode. Default "30m".
    #[serde(deserialize_with = "config_de::duration")]
    pub timeout: Duration
}

impl Default for AwayConfig {
    fn default() -> Self {
        Self {
            temp_heat: 16.0,
            temp_cool: 22.0,
            timeout: Duration::from_mins(30)
        }
    }
}

/// Backplate
///
/// ```toml
/// [backplate]
/// near_pir_threshold = 15
/// serial_port = "/dev/ttyO2"
/// wiring = { heat_wire: "W1", cool_wire: "Y1" }
/// ```
#[derive(Deserialize, Debug, Clone)]
#[serde(default)]
pub struct BackplateConfig {
    /// Minimum near proximity value to be considered as movement, default 15
    pub near_pir_threshold: u16,

    /// Path to backplate serial device file, default "/dev/ttyO2"
    pub serial_port: String,

    /// HVAC wiring configuration, default `{ heat_wire: "W1", cool_wire: "Y1" }`.
    /// Valid wire names: W1, Y1, G, OB, W2, Y2, Star.
    pub wiring: WireConfig
}

impl Default for BackplateConfig {
    fn default() -> Self {
        Self {
            near_pir_threshold: 15,
            serial_port: String::from("/dev/ttyO2"),
            wiring: WireConfig::HeatAndCool {
                heat_wire: WireId::W1,
                cool_wire: WireId::Y1
            }
        }
    }
}

#[derive(Deserialize, Debug, Clone)]
pub enum WireId {
    W1, Y1, G, OB, W2, Y2, Star
}

#[derive(Deserialize, Debug, Clone)]
#[serde(tag = "type")]
pub enum WireConfig {
    HeatAndCool {
        heat_wire: WireId,
        cool_wire: WireId
    }
}
